content/blog/2019-10-23-global-value-numbering/index.md

```diff
@@ -0,0 +1,357 @@
+This clean assumption breaks when considering multiple basic blocks.
+If `x` is assigned to in two different predecessors of the block we are currently processing, looking up the value number for any assignment involving `x` on the right hand side becomes impossible!
+
+Global value numbering sidesteps this difficulty by requiring that the input source code first be transformed to [_static single assignment (SSA)_][ssa] form. 
+In SSA, every variable name can only be assigned to once. 
+Reassignments to the same variable in the original code are translated to assignments to new variable names. 
+Because reassignments often take place in different control flow branches (to actually have the branches be useful!), SSA form needs a way to recombine different names back into a final variable. 
+SSA form relies on a special additional instruction, canonically called `phi` nodes (named to be evocative of `if`, backward) to combine variables from diverging control flow back into a single variable.
+`phi` instructions take as arguments a list of renamed variables, and assign one of the variables into a new assignment based on which control flow block was actually taken.
+
+For example, consider the following Bril code:
+
+```
+start:
+  br cond left right;
+left:
+  x : int = const 1;
+  jmp exit;
+right:
+  x : int = const 2;
+  jmp exit;
+exit:
+  print x;
+```
+
+To convert these basic blocks to Bril code, we rename variables and insert a `phi` instruction when the control flow merges again in the last block.
+
+```
+start:
+  br cond1 left right;
+left:
+  x1 : int = const 1;
+  jmp exit;
+right:
+  x2 : int = const 2;
+  jmp exit;
+exit:
+  x3 : int = phi x1 x2;
+  print x3;
+```
+
+Having unique variable assignments is enormously helpful in implementing optimization---most industrial-strength compilers, including [LLVM][] and [GCC][], rely on it internally. We followed a standard algorithm to convert Bril programs to SSA form. 
+
+The first pass inserts `phi` nodes wherever control flow merges multiple variable definitions. 
+We determine this condition by computing the [_dominance frontier_][domf] (essentially, boundaries in the control flow graph where control flow can recombine) for every basic block. 
+For every block in the dominance frontier of every variable, we insert a `phi` node with as many arguments as there are predecessors to the block. 
+One bug we hit with this part of the implementation was realizing that we only should insert `phi` nodes at the dominance frontier for variables that we assigned to multiple times. 
+For convenience of later analysis, we also augmented the base algorithm to track which basic blocks (or source) each `phi` argument originated from (for the above example, our generated `phi` would be `x3 : int = phi x1 x2 left right;`)  
```

Comment:
Huh, that's a pretty good idea… maybe I should standardize an "SSA Bril" extension that encodes phi-nodes this way.